### PR TITLE
Allow for out of order nonce

### LINKED ISSUE #2614: Tons of `Invalid Ticket senderNonce` for the past few months
There seems to be tons of `Invalid Ticket senderNonce` coming in, especially in North America:
[1] https://discord.com/channels/423160867534929930/932724294230900776/1024598556784463902
[2] https://discord.com/channels/423160867534929930/932724294230900776/1023253242697764924

It seems to be affecting all Orchestrators, did something change in the Broadcaster config? Maybe related to changes to use secondary broadcasters rather than the primary ones

### PR DESCRIPTION
**Pull request by cyberj0g**
Implementation of https://github.com/livepeer/go-livepeer/issues/2661 to fix https://github.com/livepeer/go-livepeer/issues/2614 based on introducing sorted hash map to store seen nonce values. Also added @yondonfu's test from corresponding branch, which now passes.

### CODE REVIEW

File: pm/recipient_test.go
Comment (thomshutt) [inline pm/recipient_test.go]: This is probably me misunderstanding, but I'd expect this be `.Back()` to look at the most recent one

Comment (cyberj0g) [reply pm/recipient_test.go]: Right, `Back()` would be most recent nonce, but in this test we test only one ticket anyway.

**Review by thomshutt — APPROVED** [review]
LGTM, probably good to wait for Yondon's approval too since he originally documented this issue

File: pm/recipient.go
Comment (yondonfu) [inline pm/recipient.go]: Seems simpler to just return an error if the max size is hit. If we delete the first seen nonce, then that means that in theory B could re-use that deleted nonce after the max size is hit and if we return an error if the max size is hit we at least wouldn't need to reason about this scenario. This would mean that once the max size is hit, B would *need* a new set of ticket params with a fresh `recipientRandHash`, but as mentioned in the original issue at the moment this should happen after B receives a response for each transcoded segment.

If we don't need to delete the first seen senderNonce value I think that would also simplify the data struct impl s.t. we might not need to depend on the custom ordered map package.

WDYT?

File: pm/recipient.go
Comment (yondonfu) [inline pm/recipient.go]: I don't think we need to track an expiration block per senderNonce - we only need to track an expiration block per recipientRandHash (aka `randStr` in this function). And we only need to track seen senderNonce values associated with a particular recipientRandHash until that recipientRandHash expires.

I think this would simplify the state tracking since we can track an expiration block per recipientRandHash instead of an expiration block for every senderNonce value. With the current impl I think the map ends up looking like `{ key: 1 value: X, key: 2 value: X, key 3 value: X }`.

File: pm/recipient.go
Comment (yondonfu) [inline pm/recipient.go]: If we track a single expiration block per recipientRandHash then we wouldn't need the nested loop here and instead could just clear all senderNonce values for the recipientRandHash once the expirationBlock is reached.

**Comment by Quintendos** [comment]
Ivan to fix up Yondon's comments and then ready to merge.

**Comment by cyberj0g** [comment]
Thanks for reviews @yondonfu @thomshutt, if we don't want FIFO and don't need to track expiration block number per nonce, that makes the logic simpler and removes sorted hash map dependency. I still used map instead of slice to have O(1) retrieval.

File: pm/recipient.go
Comment (yondonfu) [inline pm/recipient.go]:
```suggestion
	senderNonces, randKeySeen := r.senderNonces[randStr]
```

Nit: The plural feels a bit clearer here since the map is tracking a collection of nonces as opposed to a single one.

File: pm/recipient.go
Comment (yondonfu) [inline pm/recipient.go]: Nit: Is there an advantage of using `byte` here instead of a `bool`?

IIUC a `bool` should be 1 byte which would be the same size as a `byte` [1] so I don't think there are any storage benefits of using `byte`. If this is the case, I suggest using a `bool` as I think using true/false values will be clearer for code readability than using 0/1.

[1] Can verify using `unsafe.Sizeof()` on a `bool` var and a `byte` var.

File: pm/recipient.go
Comment (yondonfu) [inline pm/recipient.go]:
```suggestion
	if len(r.senderNonces[randStr].nonceSeen) >= maxSenderNonces {
```

Nit: IMO feels a bit clearer (no arithmetic needed) and conveys "if we are at the max size or for some reason exceeded it, return an error".

**Review by yondonfu — APPROVED** [review]
Left a few nits. Feel free to address prior to merging.

Other than that LGTM.

**Pull request merged**
